### PR TITLE
Allow any types in invokedynamic advice method signatures

### PR DESCRIPTION
**Pull request by JonasKunz**
Before this PR, instrumentations whic hare dispatched via `invokedynamic` would cause problems if they used certain types as return or paramter types:

When inserting the `invokedynamic` instruction, it is associated with a `MethodType` of the called method for the bytecode verifier to perform its job. If that `MethodType` contains references to types not visible to the instrumented class, it would make it unloadable after instrumentation.

This PR lifts this restriction by replacing all non `java.*` type references in the `MethodType` with `java.lang.Object`.
This is an implementation for [this discussion](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/11553#issuecomment-2219806452).

Unfortunately bytebuddy doesn't support this kind of transformation yet, so we have to manually erase the types via ASM before giving the advice to bytebuddy.

I'll open a feature request there to eventually allow us to do this directly with bytebuddy, in order to avoid the overhead of having to transform the advice bytecode.

### CODE REVIEW

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            }
+          }
+        };
+    classNode.accept(cv);
+    return cw.toByteArray();
+  }
+
+  private static String eraseTypes(String descriptor) {
+    Matcher matcher = TYPE_REFERENCE_PATTERN.matcher(descriptor);
+    StringBuffer result = new StringBuffer();
```
Comment (laurit) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java:83]:
```suggestion
    StringBuilder result = new StringBuilder();
```

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ * class reference are replaced with references to {@link Object}. To lookup the correct advice
+ * method, we pass the original type as string argument to the invokedynamic bootstrapping method.
+ *
+ * <p>Because bytebuddy currently doesn't support this customization, we perform the type erasure on
+ * the .class via ASM before bytebuddy parses the advice. In addition, we insert an annotation to
+ * preserve the original descriptor of the method.
+ */
+// TODO: replace this workaround when buddy natively supports altering the MethodType for
+// invokedynamic advices
+public class AdviceSignatureEraser {
```
Comment (laurit) [inline javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java:39]:
```suggestion
class AdviceSignatureEraser {
```

File: javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java (hunk 1 above)
Comment (JonasKunz) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java:83]: Unfortunately the`StringBuilder` based [Matcher methods](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/java/util/regex/Matcher.html#appendReplacement-java.lang.StringBuilder-java.lang.String-) have been added in Java 9, so we can't use them here.
However, I could do some manual fiddeling with the match start/end indices if you prefer that.

Comment (laurit) [reply javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceSignatureEraser.java:83]: Oh, that is fine then.

**Review by laurit — APPROVED** [review]

**Pull request merged**